GH-1159 Batch Insert Operations Within a Single Aggregate

ctailor2 opened the pull request:
This improves the strategy for inserting entities referenced by aggregate to perform a batch insert when there are multiple entities in a collection.

This is only possible for databases which support retrieving generated keys for batch operations through [`Statement#getGeneratedKeys`](https://docs.oracle.com/en/java/javase/17/docs/api/java.sql/java/sql/Statement.html#getGeneratedKeys()), which includes all currently supported databases other than DB2 and [MS SQL Server](https://github.com/microsoft/mssql-jdbc/issues/358). Note: DB2 _does_ support this functionality outside of the JDBC spec with [`DB2PreparedStatement#getDBGeneratedKeys`](https://www.ibm.com/docs/en/db2/11.5?topic=jdbc-db2preparedstatement-interface).

This implementation includes low level code in `BatchJdbcOperations` for performing the batch insert with a `KeyHolder` and `String[] keyColumnNames` which has largely been copied from spring framework `spring-jdbc`, pending the acceptance of a [proposal to add native support via additional `NamedParameterJdbcOperations#batchUpdate` methods](https://github.com/spring-projects/spring-framework/pull/28132). If and when this proposal is accepted and merged, `BatchJdbcOperations` can be deleted and callers into it can be replaced with calls into `NamedParameterJdbcOperations`.

Caveats:
- As part of the construction of the `AggregateChange` to be executed, inserts for a given referenced collection of entities are grouped into batches based on the presence/absence of an id value. Since this is determined only once rather than with every jdbc operation, this will produce unexpected results for any existing `BeforeSaveCallback` that is setting the value of an `@Id` annotated property. Any such operation will need to be moved to `BeforeConvertCallback` in order to continue working as expected.
- Batch insert has not been implemented for `MyBatisDataAccessStrategy` at this time. When used, this will [process the batch insert as serial inserts](https://github.com/spring-projects/spring-data-relational/pull/1191/files#diff-6eed6c7094e230d4436cca1cd6cf715d102b5f253bff63ee1ac2dde42ccb074fR165-R167).

Closes #1159

schauder commented on spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/BatchInsertStrategy.java:
```diff
@@ -0,0 +1,34 @@
+ * Strategy for executing a batch insert.
+ *
+ * @author Chirag Tailor
+ */
```
For all new classes, interfaces, methods ... we add a `@since` tag with the version introducing the artifact.
```suggestion
  * @since 2.4
 */
```

schauder commented on spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/SqlParametersFactory.java:
```diff
@@ -0,0 +1,289 @@
+	 * @param includeId an indicator of whether the insert includes the id.
+	 * @return the {@link SqlIdentifierParameterSource} for the insert. Guaranteed to not be {@code null}.
+	 */
+	<T> SqlIdentifierParameterSource getInsert(T instance, Class<T> domainType, Identifier identifier, boolean includeId) {
```
I find the name a little misleading, since the method does not returns "a Insert" which I would expect to be a SQL statement, or maybe an object wrapping such a statement. 

`forInsert` or `getParameterSourceForInsert` would be better names, I think.

ctailor2 commented on spring-data-jdbc/src/main/java/org/springframework/data/jdbc/core/convert/DataAccessStrategy.java:
```diff
@@ -35,23 +36,59 @@
-	<T> Object insert(T instance, Class<T> domainType, Identifier identifier);
+	@Deprecated
+	default <T> Object insert(T instance, Class<T> domainType, Identifier identifier) {
+		return insert(instance, domainType, identifier, false);
```
@schauder I reinstated and deprecated the original method signature for backward compatibility. Is it acceptable that the default implementation of this will no longer work for certain cases or should I try to offer more thorough concrete implementations? If needed, I could reuse the logic from [`WritingContext#idPropertyIsSet`](https://github.com/spring-projects/spring-data-relational/pull/1191/files#diff-574b599c8d79f36deb6620b56a08d08463b998baf9f246b0d1d7fc4e2a6d95b4R307-R320) in `DefaultDataAccessStrategy` to determine a more appropriate value for the `includeId` parameter.

schauder commented on spring-data-jdbc/src/test/resources/org.springframework.data.jdbc.repository/JdbcRepositoryWithListsIntegrationTests-db2.sql:
```diff
@@ -3,3 +3,20 @@ DROP TABLE dummy_entity;
 CREATE TABLE dummy_entity ( id BIGINT GENERATED BY DEFAULT AS IDENTITY ( START WITH 1 ) PRIMARY KEY, NAME VARCHAR(100));
 CREATE TABLE element (id BIGINT GENERATED BY DEFAULT AS IDENTITY (START WITH 1) PRIMARY KEY, content VARCHAR(100), Dummy_Entity_key BIGINT, dummy_entity BIGINT);
+
+CREATE TABLE root
```
We need a drop table statement for each new table. This applies to the other databases as well. 

The test fails locally, probably when run multiple times, because the database gets reused and the create table fails when a database is already present.

schauder reviewed (changes requested):
As for the old method signatures: They should behave as close to the original as possible. If they don't work for certain cases we need to identify those cases in order to decide if we can go forward with this, or if we have to move the change to the next major release.

schauder commented:
And please squash the actual feature in a single commit ,please.

ctailor2 commented:
> As for the old method signatures: They should behave as close to the original as possible. If they don't work for certain cases we need to identify those cases in order to decide if we can go forward with this, or if we have to move the change to the next major release.

I removed my previous default implementation for the deprecated `DataAccessStrategy#insert` method and added a more thorough [concrete implementation](https://github.com/spring-projects/spring-data-relational/pull/1191/files#diff-6adeda6c2cf507ae7f66b7517d5c23d0d0186adaef23e3770a36ab943e485c9cR97-R102) that works for all values of `IdValueSource`.

schauder commented:
Thanks, that's merged.
